Bug 1558572, Documented migrating masters

ahardin-rh opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1558572

ahardin-rh commented:
@adellape Do you mind taking a look and letting me know what you think of this placement?

ahardin-rh commented:
@miheer @sdodson PTAL

fcami commented:
LGTM except that "Provision 3 new virtual machines" should be "Provision as many new virtual machines as there are masters to replace" - @sdodson ?

ahardin-rh commented:
@openshift/team-documentation PTAL

mburke5678 commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
+*_/etc/origin/master/ca.serial.txt_* from the *_/etc/origin/master_* directory
+to the new master host that is listed first in your hosts file. This directory
+is on the old master host that was listed first.
+
```
Do we need the full paths to the files and the path to the parent file?
 Copy the *_ca.crt_*, *_ca.key_*, and *_ca.serial.txt_* files from the *_/etc/origin/master_* directory to the new master host...  OR
 Copy *_/etc/origin/master/ca.crt_*, *_/etc/origin/master/ca.key_*, and *_/etc/origin/master/ca.serial.txt_* to the new master host...

mburke5678 replied:
The last sentence is a bit confusing. Where is the hosts file?
_Copy (the files) from the old master host to the new master host. The old and new master to use is the first master listed in the *_hoists_* file of the respective masters._

miheer replied:
You can say inventory file rather than hosts file. Then say by default it is /etc/ansible/hosts

miheer replied:
please give the full paths

mburke5678 commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
+
+.. Take the IDs from the the command above and remove the old members using the
+IDs:
++
```
s/from the the command above/from the output of the command above ??

mburke5678 commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
+
+.. Mark the current node (in the old data center)
+xref:../admin_guide/manage_nodes.adoc#marking-nodes-as-unschedulable-or-schedulable[unscheduled].
+
```
@kalexand-rh suggests not using parenthesis ??

mburke5678 commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
+. Add or expand the cluster. If you want to add 3 masters with etcd collated,
+ repeat this section 3 times.
+
+.. Add a master using the xref:adding-nodes-advanced[master *_scaleup.yml_*
```
Maybe make parallel with the ectd step? Like:
Add a xref:adding-nodes-advanced[master]. In step 3 of that process, add the host of the
new data center in `[new_masters]` and `[new_nodes] and run the master *_scaleup.yml_* playbook.

mburke5678 commented:
@ahardin-rh Couple of comments. Otherwise /lgtm

miheer commented:
@ahardin-rh the doc looks good to me

miheer commented:
@ahardin-rh just replied to some comments by @mburke5678

adellape commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
+and Restore] topic.
+====
+
+. Provision as many new virtual machines (VMs) as there are masters to replace.
```
This could be just "systems" or something more generic than VMs?

ganhuang commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
 xref:../install_config/install/advanced_install.adoc#advanced-verifying-the-installation[verify the installation].
+
+[[replacing-existing-masters]]
+== Replacing Existing Masters
```
After reading the whole section, it appears to me it should be the case of replacing existing masters **collocated with etcd**.

ganhuang commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
+. Provision as many new virtual machines (VMs) as there are masters to replace.
+
+. Add or expand the cluster. If you want to add 3 masters with etcd collated,
+ repeat this section 3 times.
```
For the master or etcd scaling up, why not scale up 3 masters or 3 etcd in a bunch directly.

ganhuang commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
+.. Copy *_/etc/origin/master/ca.crt_*, *_/etc/origin/master/ca.key_*, and
+*_/etc/origin/master/ca.serial.txt_* from the *_/etc/origin/master_* directory
+to the new master host that is listed first in your hosts file. This directory
+is on the old master host that was listed first.
```
In my test, we only need to copy `/etc/origin/master/ca.serial.txt`. The rest of the files are exactly the same.

ganhuang commented:
One more thing, we also need to remove master node from the cluster.

ganhuang commented:
We'd better put the step "Shut down the services" in the end in case of causing downtime while updating "master-config.yaml" while performing step "Remove the old etcd clients from the master-config.yaml file".

ganhuang commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,133 @@ $ ansible-playbook [-i /path/to/file] \
+.. Shut down the services:
++
+----
+# systemctl stop atomic-openshift-master-api
```
master includes api and controllers services.

ahardin-rh commented:
@sdodson @fcami @miheer I think we're getting closer. I made some edits, but looks like we still need to address:

- Why would users go through this process? What are some likely scenarios?
- We need to add step(s) to remove the master node from the cluster and the old masters from the loadbalancer. What are the steps for this and when should this happen in the process?

Updated preview build: http://file.rdu.redhat.com/%7Eahardin/04052018/migrate-masters/install_config/adding_hosts_to_existing_cluster.html#replacing-existing-masters

Thanks!

miheer commented:
The users will have to go through this process when they are migrating the vm's to a different data center where the network will change and thereby the ip's assigned to them.

From my understanding the removal of the masters from the haproxy config which by default installed as load balancer during native install can be done at the end i.e after stopping the api and controller service.  Anyway when the master service is turned off it won't be considered. But again some customers use f5 load balancer so I don't have much idea how we can remove it from there. So may be @sdodson  might help us telling what needs to be removed from the haproxy config used in native install.

Also if we are about to decommission the vm itself then we should also stop atomic-openshift-node service on the master to be removed and then delete the node resource using oc delete node.

Also here in this point, we may need to specify that value of  ETCDCTL_API depends on the version being used.

Remove the old etcd members from the cluster.
# ETCDCTL_API=2

ahardin-rh commented:
@miheer Thanks! Updated with the latest information:
 http://file.rdu.redhat.com/%7Eahardin/04052018/migrate-masters/install_config/adding_hosts_to_existing_cluster.html#replacing-existing-masters

Getting closer.

@sdodson PTAL, noting the comment above. Thanks!

sdodson commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,152 @@ $ ansible-playbook [-i /path/to/file] \
+----
+# ETCDCTL_API=2
+# source /etc/etcd/etcd.conf
+# etcdctl --cert-file=$ETCD_PEER_CERT_FILE --key-file=$ETCD_PEER_KEY_FILE \
```
Move the ETCDCTL_API=2 to the same line as the etcdctl command so that it doesn't persist. example
`# ETCDCTL_API=2 etcdctl --cert-file=$ETCD_PEER_CERT_FILE --key-file=$ETCD_PEER_KEY_FILE \`
...

sdodson commented on install_config/adding_hosts_to_existing_cluster.adoc:
```diff
@@ -309,3 +309,152 @@ $ ansible-playbook [-i /path/to/file] \
+----
+# ETCDCTL_API=2
+# source /etc/etcd/etcd.conf
+# etcdctl --cert-file=$ETCD_PEER_CERT_FILE --key-file=$ETCD_PEER_KEY_FILE \
```
move ETCDCTL_API=2 to the beginning of this command too.

ahardin-rh commented:
Thanks! This is now updated. Updated preview build: http://file.rdu.redhat.com/%7Eahardin/04052018/migrate-masters/install_config/adding_hosts_to_existing_cluster.html#replacing-existing-masters

ahardin-rh commented:
QE verified in the BZ

ahardin-rh commented:
/cherrypick enterprise-3.7

ahardin-rh commented:
/cherrypick enterprise-3.9

ahardin-rh commented:
/cherrypick enterprise-3.10
